[AMDGPU][Scheduler] Refactor VGPR rematerialization during scheduling

lucas-rami opened the pull request:
AMDGPU scheduler's `PreRARematStage` attempts to increase function occupancy w.r.t. VGPR usage by rematerializing trivial VGPR-defining instruction next to their single use. It first collects all eligible trivially rematerializable instructions in the function, then sinks them one-by-one while recomputing occupancy in all affected regions each time to determine if and when it has managed to increase overall occupancy. If it does, changes are committed to the scheduler's state; otherwise modifications to the IR are reverted and the scheduling stage gives up. In both cases, this scheduling stage currently involves repeated queries for up-to-date occupancy estimates and some state copying to enable reversal of sinking decisions when occupancy is revealed not to increase. The current implementation also does not accurately track register pressure changes in all regions affected by sinking decisions.

This commit refactors this scheduling stage, improving RP tracking and splitting the stage into two distinct steps to avoid repeated occupancy queries and IR/state rollbacks.

1. Analysis and collection (`canIncreaseOccupancy`). The number of VGPRs to save to increase occupancy in each region at minimum occupancy is computed. Then, instructions eligible for rematerialization are collected, stopping as soon as enough have been identified to be able to increase function occupancy (according to slightly optimistic heuristics). If there aren't enough of such instructions, the scheduling stage stops here.
2. Rematerialization (`sinkTriviallyRematInsts`). Instructions collected in the first step are rematerialized one-by-one. Now we are able to directly update the scheduler's state since we have already done the occupancy analysis and know we won't have to rollback any state. Register pressures for impacted regions are recomputed only once, as opposed to at every sinking decision.

This is the first change in a series of changes to improve rematerialization during scheduling. The next immediate steps are to add support for SGPR/AGPR rematerialization, then potentially relax some constraints on instructions we consider eligible for rematerialization.

lucas-rami committed:
Working refactoring of simple VGPR remat.

lucas-rami committed:
Edit stale comments and slightly rearch stage

lucas-rami committed:
Wrap comment correctly

lucas-rami committed:
Fix format

arsenm commented on llvm/test/CodeGen/AMDGPU/machine-scheduler-sink-trivial-remats.mir:
```diff
@@ -5913,4 +5913,141 @@ body:             |
+  bb.3:
+    S_NOP 0, implicit %0, implicit %1
+    S_ENDPGM 0
+...
```
Can you test rematerializing over an exec modification? Also the undef subreg def case

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
+      // non-subregister) operand.
+      MachineInstr &DefMI = *MI;
+      if (DefMI.isBundle() || !DefMI.getNumOperands() ||
+          !DefMI.getOperand(0).isReg() || DefMI.getOperand(0).getSubReg())
```
Rematerializing undef subregister defs should work. Also does this need to verify that operand 0 is a def? But this is repeating isReallyTriviallyReMaterializeable's logic anyway, most of these pre-checks should be redundant

lucas-rami replied:
Thanks for the catches. We indeed don't need to check that operand 0 is a def here as it is already implied by `isTriviallyReMaterializable`. I now allow undef subregister defs as well (see `test_occ_7_sink_one_def_of_undef_subreg_for_8`).

However, I am now forced to do something that seems slightly weird to me during rematerialization; I have to explicitly set the subreg on the new MI's def after calling `TII.reMaterialize` even though the subreg is passed to the method already (if I don't do this I hit an assert in `DAG.LIS->createAndComputeVirtRegInterval(Reg)` at the end). Not sure this is the correct way to do this.

arsenm replied:
The subregister should survive the clone and substitute. What assert?

lucas-rami replied:
[This one in `LiveInterval::computeSubRangeUndefs`](https://github.com/llvm/llvm-project/blob/00b07aa399b870d4705c4fb628e25d62c8ccc27b/llvm/lib/CodeGen/LiveInterval.cpp#L971). Looking at the in-flight MIR before explicitly setting the subreg after rematerialization it seems that the `NewMI`'s def reg is being assigned a different non-intersecting subregister than the original one, which leads to the two masks tested here to be completely unrelated.

In the `test_occ_7_sink_one_def_of_undef_subreg_for_8` test for example, after calling `SIInstrInfo::reMaterialize` but before setting the subreg explicitly, the new MI uses `sub2` instead of `sub1`.

```
[PreRARemat] Rematerializing undef %32.sub1:vreg_64_align2 = V_MOV_B32_e32 23, implicit $exec
[PreRARemat] New MI:         undef %32.sub2:vreg_64_align2 = V_MOV_B32_e32 23, implicit $exec
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
+        // instruction and the end of the region.
+        RA_DEBUG(dbgs() << "  Instruction's defining region is optimizable: ");
+        RematUseful = true;
+        auto RegMask = DAG.RegionLiveOuts.getLiveRegsForRegionIdx(I)[Reg];
```
No auto

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
-      break;
+void PreRARematStage::sinkTriviallyRematInsts(
+    ArrayRef<RematInstruction> RematInstructions, const GCNSubtarget &ST,
+    const TargetInstrInfo *TII) {
```
SIInstrInfo

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.h:
```diff
@@ -419,29 +420,48 @@ class ClusteredLowOccStage : public GCNSchedStage {
+    MachineInstr *RematMI;
+    /// Region containing the rematerializable instruction.
+    unsigned DefRegion;
+    /// Single use of the rematerializable instruction's defined register,
+    /// located in a different block.
+    MachineInstr *UseMI;
```
AFAIU it is per region, and in the case of a rematerialization both the region where the instruction was rematerialized from and the one it is rematerialized to will be rescheduled after this scheduling stage takes place.

In the future we could improve rematerialization to work even if the def and use are in the same block but not the same region, which from what I understand can happen and make sense.

lucas-rami committed:
Address review comments

lucas-rami committed:
Remove useless closure argument

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
-    }
+      // We only care to rematerialize the instruction if its single use is in a
+      // different block.
+      MachineInstr *UseMI = &*DAG.MRI.use_instr_nodbg_begin(Reg);
```
I added `MachineRegisterInfo::getOneNonDBGUser` for this purpose.

lucas-rami committed:
Fix format

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1671,6 +1667,20 @@ void PreRARematStage::sinkTriviallyRematInsts(
+    Register Reg = NewMI->getOperand(0).getReg();
+    DAG.LIS->RemoveMachineInstrFromMaps(*OldMI);
+    OldMI->eraseFromParent();
+    DAG.LIS->removeInterval(Reg);
+    DAG.LIS->createAndComputeVirtRegInterval(Reg);
```
Is it possible to spot fix the interval instead of computing from scratch

lucas-rami replied:
Should be possible, working on this (might take me a minute since I am not so familiar with `LiveInterval` and related classes yet).

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1716,6 +1712,11 @@ bool PreRARematStage::isTriviallyReMaterializable(const MachineInstr &MI) {
   if (!DAG.TII->isTriviallyReMaterializable(MI))
     return false;
 
+  // Even though TargetInstrInfo::isReallyTriviallyReMaterializable already
+  // ensures that the instruction has no virtual register uses,
+  // SIInstrInfo::isReallyTriviallyReMaterializable may consider an instruction
+  // rematerializable and return before calling its parent's method, so we need
+  // to double-check here.
```
Seems like a bug? All of these rematerialize functions are confusing

lucas-rami replied:
It is a bit confusing to me as well. Our `SIInstrInfo` override considers that rematerializing VOP{1,2,3} instructions with virtual register uses is trivial, which I don't really agree with. Removing the override in favor of its parent's method breaks many unit tests however, so if that's ok I will leave this as is for now and investigate this further in the future.

arsenm replied:
I think 08d7eec06e8cf5c15a96ce11f311f1480291a441 never got finished / reapplied

lucas-rami committed:
Update live interval manually instead of fully recomputing

lucas-rami committed:
Remove spurious live interval recomputing

lucas-rami committed:
Early return on SGPR-limited region

lucas-rami commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,231 +1469,246 @@ void GCNSchedStage::revertScheduling() {
-    if (Def->getParent() == UseI->getParent())
+    // We do not rematerialize SGPR-defining regions yet so do not bother
+    // optimizing regions whose occupancy is SGPR-limited.
+    if (ST.getOccupancyWithNumSGPRs(RP.getSGPRNum()) == DAG.MinOccupancy)
```
Thanks for the catch!

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
-        // searching later when updating RP.
-        RematDefToLiveInRegions[Def].push_back(I);
-      }
+    // We do not rematerialize SGPR-defining regions yet so do not bother
```
rematerialize SGPR-defining regions -> rematerialize SGPR-defining regions instructions

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
+    // Remove rematerialized instruction from BBLiveInMap since we are sinking
+    // it from its MBB. Also remove it from live intervals and from its MBB.
+    MachineInstr *OldMI = Remat.RematMI;
     DAG.BBLiveInMap.erase(OldMI);
```
I don't think this is maintaining the map the way we want.

lucas-rami replied:
After taking a closer look, it's probably not indeed (and potentially the same applies to `MBBLiveIns`). Working to fix this.

lucas-rami replied:
Spent some more time looking at this, and it seems both `BBLiveInMap` and `MBBLiveIns` are not accessed after sinking instructions. Their only uses are in `computeBlockPressure` which is only called for other scheduling stages.

However, as I was trying to understand the exact purpose of `BBLiveInMap` I noticed that while `GCNScheduleDAGMILive::getRegionLiveInMap` seems to want to compute the live reg set at the beginning of the initial region of each block, the `BB` pointer is never updated. Seems like it should be inside the outer do/while loop.

https://github.com/llvm/llvm-project/blob/6780ab371ff8ce5e2b08b5fddd75a8296d3debdd/llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp#L886-L894

What happens right now is that non-initial regions (if any) in the first block are skipped but they are not in other blocks. I don't see why the first block should be treated differently, is that a bug?

jrbyrnes replied:
> Spent some more time looking at this, and it seems both BBLiveInMap and MBBLiveIns are not accessed after sinking instructions.

We should probably either completely abandon trying to maintain these, or do it correctly. 

> What happens right now is that non-initial regions (if any) in the first block are skipped 

Yes I would expect this to build of region identifiers to the LiveRegSet for each region -- similar to getRegionLiveOutMap. 

However, we currently only use the LiveRegMap to get the LiveRegs per block and not per region, so we aren't exposed to any undue RP inaccuracies from this currently

lucas-rami replied:
Sounds good, I removed the updates to `BBLiveInMap` and `MBBLiveIns` from the stage for now since they are not accessed after it anyway. I also fixed the issue in `getRegionLiveInMap`.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
+      if (Remat.UseMI->getParent() != DAG.Regions[I].first->getParent()) {
+        // Register is live-through and not used in this block.
+        LaneBitmask PrevMask = RegionLiveIns[Reg];
+        DAG.Pressure[I].inc(Reg, PrevMask, LaneBitmask::getNone(), DAG.MRI);
```
Should we be using the LaneMask?

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
+    SlotIndex NewDef = Slots->getInstructionIndex(*NewMI).getRegSlot();
+    SlotIndex NewKill = Slots->getInstructionIndex(*Remat.UseMI).getRegSlot();
+
+    auto UpdateSegment = [&](LiveRange::Segments &Segments) -> void {
```
Can we use the LIS for this?

lucas-rami replied:
Do you mean for this particular segment change? If so I don't know if there exists some function that does this already. We would just like to avoid doing a more expensive full recomputation of the LI with

```cpp
DAG.LIS->removeInterval(Reg);
DAG.LIS->createAndComputeVirtRegInterval(Reg);
```

jrbyrnes reviewed (commented):
The analysis doesn't prove that we will change occupancy through rematerialization -- I think this is a complex problem involving the live regs at the rematerialization point, which may change with scheduling. Because of this, we may want to keep the rollback functionality?

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
-    NewPressure[Idx] = DAG.Pressure[Idx];
-    NewLiveIns[Idx] = DAG.LiveIns[Idx];
+    unsigned NumVGPRs = RP.getVGPRNum(ST.hasGFX90AInsts());
+    unsigned NumToIncreaseOcc = ST.getNumVGPRsToIncreaseOccupancy(NumVGPRs);
```
What about the case to eliminate spilling? For spilling in unified register file case (hasGFX90AInsts), there are additional constraints we need to consider (see https://github.com/llvm/llvm-project/pull/85860)

lucas-rami replied:
At this point I am not considering the trade-off between occupancy and spilling i.e., I am not allowing any spilling while trying to increase occupancy. In that context, is there anything else I should consider? I plan to properly model excess ArchVGPR/AccVGPR in both unified and non-unified RFs in a following PR which will be able to remat instructions for all classes of virtual def regs.

jrbyrnes replied:
Should we be trying to increase occupancy in cases where there is spilling ? Or just eliminate the spilling ?  Eliminating spilling with a reduction in ArchVGPR usage alone may not be possible if AccVGPR pressure is too high.

lucas-rami replied:
> Should we be trying to increase occupancy in cases where there is spilling ? Or just eliminate the spilling ?

Thanks for the suggestion, I had not explicitly considered trying to reduce/remove spilling even if we are unable to increase occupancy indeed. I have added support for it. In situation where occupancy is 1 and there is spilling, the stage now tries to collect as many rematerializable instructions as necessary to reduce/eliminate spilling, and will rematerialize just enough of them to eliminate spilling (or at least reduce it) even if we cannot increase occupancy (see new unit test `test_occ_1_sink_for_spill`). If it is possible to eliminate spilling *and* increase occupancy then it rematerializes just enough instructions for both of them to happen. 

> Eliminating spilling with a reduction in ArchVGPR usage alone may not be possible if AccVGPR pressure is too high.

For now I still use the same approximation as in the current implementation, and plan to properly model excess RP for all register types in a further PR if that is ok.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,238 +1470,277 @@ void GCNSchedStage::revertScheduling() {
-    if (ST.getOccupancyWithNumSGPRs(SGPRUsage) == DAG.MinOccupancy)
-      break;
+      // We only support rematerializing virtual VGPRs with one definition.
+      Register Reg = DefMI.getOperand(0).getReg();
+      if (!Reg.isVirtual() || !DAG.LIS->hasInterval(Reg) ||
+          !SRI->isVGPRClass(DAG.MRI.getRegClass(Reg)) ||
+          !DAG.MRI.hasOneDef(Reg))
+        continue;
```
Most of this is covered by isTriviallyReMaterializable, I think. !hasInterval would be malformed?

lucas-rami replied:
If `TargetInstrInfo::isTriviallyReMaterializable` early returns on

```cpp
MI.getOpcode() == TargetOpcode::IMPLICIT_DEF && MI.getNumOperands() == 1
```

then I think I need the "is virtual" check. And I am not sure that `TargetInstrInfo::isReallyTriviallyReMaterializable` implies that the register has only one def.

I left the `hasInterval` check because it was here before so I assumed the interval could be missing, but perhaps not.

lucas-rami commented:
> The analysis doesn't prove that we will change occupancy through rematerialization -- I think this is a complex problem involving the live regs at the rematerialization point, which may change with scheduling. Because of this, we may want to keep the rollback functionality?

I completely agree, but my rationale for removing the rollback functionality is that after rescheduling `finalizeGCNRegion` -> `checkScheduling` -> `revertScheduling` can already revert scheduling on a per-region basis if we caused a drop in occupancy or extra spilling to happen, at the cost of a hit to compile-time compared to rollbacking immediately. I also suppose that rescheduling post-remat can improve occupancy/spilling beyond what we can assess at the end of rematerialization. Doing an immediate rollback would make us miss on those potential improvements I think.

jrbyrnes commented:
> > The analysis doesn't prove that we will change occupancy through rematerialization -- I think this is a complex problem involving the live regs at the rematerialization point, which may change with scheduling. Because of this, we may want to keep the rollback functionality?
> 
> I completely agree, but my rationale for removing the rollback functionality is that after rescheduling `finalizeGCNRegion` -> `checkScheduling` -> `revertScheduling` can already revert scheduling on a per-region basis if we caused a drop in occupancy or extra spilling to happen, at the cost of a hit to compile-time compared to rollbacking immediately. I also suppose that rescheduling post-remat can improve occupancy/spilling beyond what we can assess at the end of rematerialization. Doing an immediate rollback would make us miss on those potential improvements I think.

revertScheduling doesn't un-materialize the remats. 

It's possible that we satisfy the heuristic via rematerialization, but don't increase occupancy before rescheduling. Post-remat rescheduling may improve occupancy / spilling, but it also may not; I think we should err on the side of not doing the remats as this may introduce regressions.

If we have reason to believe that we should always try to reschedule after remat, then we should move the remat rollback functionality into revertScheduling

lucas-rami committed:
Incorporated feedback and fixed pre-existing bugs

- The stage now tries to remove any VGPR spilling before trying to
  increase occupancy. In all cases it remats just enough instructions to
  reduce/eliminate spilling and/or increase occupancy.
- Stop updating MBBLiveIns and BBLiveInMap in the stage, since these are
  not accessed.
- Fixed region boundary update logic, it did not work when a region
  became empty due to remating all of its MIs.
- Fixed issue in getRegionLiveMap wherein the BB pointer was not updated
  correctly.

lucas-rami committed:
[NFC] Tweak some comments

lucas-rami commented:
> revertScheduling doesn't un-materialize the remats.

I did not realize this, thanks. I see the need for potentially rollbacking our decisions.

> If we have reason to believe that we should always try to reschedule after remat, then we should move the remat rollback functionality into revertScheduling

I think we should always try to reschedule since the remat heuristics are only "slightly optimistic". In particular, they are exact in regions where rematerialized registers are simply live-through. My motivating example for this series of changes is a case of LICM bringing many defs inside the pre-header of a grid stripe loop containing the whole kernel (which I think is a fairly common situation), causing a lot of spilling. Most affected regions in such cases are those where rematable registers are live-through.

I am working on adding post-scheduling rollback functionality to the remat stage and trying to minimize the performance impact in the happy path where no rollbacking is required.

lucas-rami committed:
Allow rematerializations to be rollbacked

jrbyrnes commented on llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp:
```diff
@@ -1185,6 +1185,28 @@ unsigned getNumWavesPerEUWithNumVGPRs(unsigned NumVGPRs, unsigned Granule,
+unsigned getVGPRReductionToEliminateSpilling(const MCSubtargetInfo *STI,
+                                             unsigned NumVGPRs) {
+  unsigned Granule = getVGPRAllocGranule(STI);
+  unsigned TotalNumVGPRs = getTotalNumVGPRs(STI);
```
This returns 512 for the unified register file case when we actually want 256 since we only care about the arch VGPR pressure

lucas-rami replied:
I introduced a family of functions to be able to query these numbers for ArchVGPRs only.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,284 +1488,460 @@ void GCNSchedStage::revertScheduling() {
+        // We have eliminated spilling in the region.
+        LLVM_DEBUG(dbgs() << "sinking eliminates spilling in region " << I
+                          << '\n');
+        if (--NumRegionsSpilling)
```
Should it be `if (!--NumRegionsSpilling)` -- that is, the limit is set when we eliminate all regions with spilling?

lucas-rami replied:
Yep, that's a typo, thanks.

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,284 +1488,460 @@ void GCNSchedStage::revertScheduling() {
+        // We have eliminated spilling in the region.
+        LLVM_DEBUG(dbgs() << "sinking eliminates spilling in region " << I
+                          << '\n');
+        if (--NumRegionsSpilling)
```
can we just return true if we've eliminated spilling? I'm not sure we should be trying to eliminate spilling and increase occupancy

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,284 +1488,460 @@ void GCNSchedStage::revertScheduling() {
+      // using region, maximum RP may or may not decrease, so we will mark it
+      // for re-computation after all materializations have taken place.
+      RegionLiveIns.erase(Reg);
+      LaneBitmask PrevMask = RegionLiveIns[Reg];
```
Accessing the element after erasing?

lucas-rami replied:
Thanks for the catch

jrbyrnes commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1467,284 +1488,460 @@ void GCNSchedStage::revertScheduling() {
+void PreRARematStage::finalizeGCNSchedStage() {
+  if (AchievedOcc == TargetOcc) {
+    // We achieved the target occupancy through rematerializations alone, but
+    // rescheduling after lowered occupancy. Scheduling has already been
```
`finalizeGCNSchedStage` is always called and `AchievedOcc` and `TargetOcc` were only set during `initGCNSchedStage` -- I don't see how this condition implies that rescheduling has lowered occupancy. Maybe these fields were meant to be updated after we schedule each region?

lucas-rami replied:
This was a stale comment indeed, I updated the comment/condition.

lucas-rami commented:
Closing because of seemingly infinite processing, replaced by #125885.
